File: 23-merge-k-sorted-lists/merge-k-sorted-lists.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode* mergeKLists(struct ListNode** lists, int listsSize) {
    struct ListNode merged;
    merged.next = NULL;

    for (int i = 0; i < listsSize; i++) {
        struct ListNode* temp = lists[i];

        while (temp != NULL) {
            struct ListNode* node = (struct ListNode*)malloc(sizeof(struct ListNode));
            node->val = temp->val;
            node->next = NULL;

            struct ListNode* prev = &merged;
            while (prev->next != NULL && prev->next->val < node->val) {
                prev = prev->next;
            }

            node->next = prev->next;
            prev->next = node;

            temp = temp->next;
        }
    }

    return merged.next;
}

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        
        ListNode merged = new ListNode(0);

        for(ListNode curr : lists) {
            ListNode temp = curr;
            
            while(temp != null) {
                ListNode node = new ListNode(temp.val);
                ListNode prev = merged;
                
                while (prev.next != null && prev.next.val < node.val) {
                    prev = prev.next;
                }
                node.next = prev.next;
                prev.next = node;
                temp = temp.next;
            }
        }
        return merged.next;
    }
}